Compute accurate dependencies for all the backends on any system

dra27 opened the pull request:
The primary goal of this PR is to allow `make depend` to be run on any architecture. In particular, it means that `make alldepend` can be run on macOS arm64 hardware, without then failing CI.

As a side-effect, it also means that non-x86 architectures no longer unnecessarily (re-)build the x86-specific backend modules.

The root `.depend` file now contains a section for each architecture, selected using `$(ARCH)`. The construction is a little bit fiddly, but hopefully well-enough commented in the code.

dra27 committed:
Parallelise alldepend target

dra27 committed:
Remove architecture-specific files from .depend

dra27 committed:
Include the dependencies for all the emitters

dra27 committed:
Don't invalidate ocamlopt with make alldepend

tmcgilchrist reviewed (approved):
I've been test driving this change on macOS/ARM64 and it has been working nicely. 
It also avoids the issue with making changes on other platforms and then needing to regenerate `.depend` on an x86_64 platform.

dustanddreams commented on Makefile:
```diff
@@ -2474,18 +2474,72 @@ partialclean::
+	@$$(MAKE) ARCH=$(1) asmcomp/emit.ml $$(ARCH_SPECIFIC)
+	@$$(OCAMLDEP) $$(OC_OCAMLDEPFLAGS) -I asmcomp $$(INCLUDES) \
+	  $$(OCAMLDEPFLAGS) asmcomp/emit.ml $$(ARCH_SPECIFIC) >> asmcomp.$(1).depend
+	@echo 'endif # ifeq $$$$(ARCH) = "$(1)"' >> asmcomp.$(1).depend
```
That should either be
```suggestion
	@echo 'endif # ifeq $$$$(ARCH) "$(1)"' >> asmcomp.$(1).depend
```
to match the actual syntax, or
```suggestion
	@echo 'endif # if $$$$(ARCH) == "$(1)"' >> asmcomp.$(1).depend
```
to be human-friendly, IMHO.

dra27 replied:
Nice catch, thanks! I agree that the second way is more friendly, but I've left it in the style of the first, just as that's at least consistent with the other Makefile fragments.

dustanddreams reviewed (approved):
That's a very nice idea!

shindere commented:
Many thanks for this beautiful and useful piece of owrk!

Here are a few notes:

A suggestion for the first commit. How about introducing:

```
DEP_FILES = $(addsuffix .depend, $(DEP_DIRS))
```

and you could then rewrite:

```
.INTERMEDIATE: $(DEP_FILES)
```

and

```
depend: $(DEP_FILES) | beforedepend
```

Re: the second commit, perhaps the commit message could be clarified to
explain that the commit does not _just_ remove the dependencies but rather
that it moves them form .depend to asmcomp.depend, or something along these
lines?

In the third commit, a very minor question about the name of the files
stiring the dependencies for arch-specific code. Wouln't be more coherent
withwhat is done elsewhere to call theme `asmcom/ARCH.depend` rather than
`asmcom.ARCH.depend`?

Finally, many thanks for having taken care to not invalidate the native
compiler. In thelong run, I hope we can come up with a solution which will
be nicer than having tomove files around but I am fully ocnvince that, in
the present situation, what has been proposed is perfectly legitimate and
the best tihng one could come up with.

Thanks a lot, too, for the very nice documentation comment you have added to
the Makefile. I think the explanation is clear even for dummies of my kind.
I have one very minor suggestion but please feel free, as a native-speakier,
to either take it into account or ignore it as seems most appropriate to
you.

The current comment ends with: `of make therefore doesn't rebuild anything (correctly).`
I think I'd replace correctly by `rightfully` and put it between before and
doesn't, so it owuld become:
`of make therefore rightfully doesn't rebuild anything.`

(Things would feel easier to read to me that way.)

dra27 commented:
All done, hopefully, thanks @shindere:
- Added `$(DEP_FILES_)`
- The commit message on the second commit is correct as it stands - the dependencies are removed from asmcomp.depend which therefore means they no longer appear in .depend
- Changed the location of the arch-specific intermediate files in the third commit (and removed the unnecessary .INTERMEDIATE for them)
- I moved the word _correctly_ in the comment in the fourth commit, but _rightfully_ is not the same as _correctly_.

shindere commented:
All good, thanks! Merging.

gasche commented:
A micro point: now there is a visual mismatch between the general build output look, with stylized commands such as `OCAMLDEP middle_end.depend`, and the new `cd asmcomp; ln -sf riscv/arch.mli .` steps that show up in the build output:

```
make[1]: Entering directory '/home/gasche/Prog/ocaml/github-matching'
  GEN asmcomp/emit.ml
cd asmcomp; ln -sf amd64/arch.mli .
cd asmcomp; ln -sf amd64/arch.ml .
cd asmcomp; ln -sf amd64/proc.ml .
cd asmcomp; ln -sf amd64/CSE.ml .
cd asmcomp; ln -sf amd64/selection.ml .
cd asmcomp; ln -sf amd64/scheduling.ml .
cd asmcomp; ln -sf amd64/reload.ml .
cd asmcomp; ln -sf amd64/stackframe.ml .
make[1]: Leaving directory '/home/gasche/Prog/ocaml/github-matching'
make[1]: Entering directory '/home/gasche/Prog/ocaml/github-matching'
  GEN asmcomp/emit.ml
cd asmcomp; ln -sf arm64/arch.mli .
cd asmcomp; ln -sf arm64/arch.ml .
cd asmcomp; ln -sf arm64/proc.ml .
cd asmcomp; ln -sf arm64/CSE.ml .
cd asmcomp; ln -sf arm64/selection.ml .
cd asmcomp; ln -sf arm64/scheduling.ml .
cd asmcomp; ln -sf arm64/reload.ml .
cd asmcomp; ln -sf arm64/stackframe.ml .
[...]
```

I think that these steps could be hidden from the output, or summarized by a single stylized command `LINK_BACKEND arm64`.

gasche commented:
@nojb silenced those `ln` commands in #13349. Thanks!
